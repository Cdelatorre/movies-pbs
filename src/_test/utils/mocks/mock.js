export const moviesData = {
  viewedMovies: [
    {
      id: "1",
      title: "Movie 1",
      genres: ["Horror", "Comedy"],
      viewed: true,
      img: "https://picsum.photos/id/118/300/400",
      createdAt: "2022-10-16T20:25:58.420Z",
    },
  ],
  unViewedMovies: [
    {
      id: "2",
      title: "Movie 2",
      genres: ["Horror", "Comedy"],
      viewed: false,
      img: "https://picsum.photos/id/159/300/400",
      createdAt: "2022-10-16T20:51:29.081Z",
    },
    {
      id: "3",
      title: "Movie 3",
      genres: ["Romance"],
      viewed: false,
      img: "https://picsum.photos/id/122/300/400",
      createdAt: "2022-10-16T20:51:45.249Z",
    },
    {
      id: "4",
      title: "Movie 4",
      genres: ["Test"],
      viewed: false,
      img: "https://picsum.photos/id/168/300/400",
      createdAt: "2022-10-16T20:52:14.784Z",
    },
    {
      id: "5",
      title: "Movie 5",
      genres: ["Random"],
      viewed: false,
      img: "https://picsum.photos/id/149/300/400",
      createdAt: "2022-10-16T20:52:36.529Z",
    },
  ],
};

export const singleMovieData = {
  id: "1",
  title: "Movie 1",
  genres: ["Horror", "Comedy", "Romance"],
  viewed: false,
  img: "https://picsum.photos/id/149/300/400",
  createdAt: "2022-10-16T20:52:36.529Z",
};

export const activeFiltersData = {
  activeFilters: [],
  search: "",
};

export const activeLoadingData = {
  value: false,
};

export const emptyStoreData = {
  movies: { viewedMovies: [], unViewedMovies: [] },
  filters: { activeFilters: [], search: "" },
  loading: { value: false },
};

export const sampleFiltersData = ["Horror", "Comedy", "Romance"];

export const customStoreData = (movies, activeFilters, loading) => {
  return {
    movies: movies ? moviesData : emptyStoreData.movies,
    filters: activeFilters ? activeFiltersData : emptyStoreData.filters,
    loading: loading ? activeLoadingData : emptyStoreData.loading,
  };
};
