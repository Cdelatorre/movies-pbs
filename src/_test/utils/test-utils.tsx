import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { AppStore, RootState, setupStore } from "../../store";
import { BrowserRouter, Router } from "react-router-dom";
import { emptyStoreData } from "./mocks/mock";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = emptyStoreData,

    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  route?: string
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (route) window.history.pushState({}, "Test routing", route);

    return (
      <BrowserRouter>
        <Provider store={store}>{children}</Provider>;
      </BrowserRouter>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithRouter(ui: React.ReactElement, history?: any) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (history) {
      return (
        <Router location="/" navigator={history}>
          {children}
        </Router>
      );
    } else {
      return <BrowserRouter>{children}</BrowserRouter>;
    }
  }
  return { ...render(ui, { wrapper: Wrapper }) };
}
