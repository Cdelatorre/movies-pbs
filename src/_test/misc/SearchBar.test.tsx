import { fireEvent, screen } from "@testing-library/react";
import SearchBar from "../../components/misc/search-bar";
import { customStoreData } from "../utils/mocks/mock";
import { renderWithProviders } from "../utils/test-utils";

describe("Renders SearchBar component correctly", () => {
  test("Renders SearchBar default component correctly", () => {
    renderWithProviders(<SearchBar />);

    const searchBar = screen.getByTestId("search-bar");
    const input = screen.getByTestId("input-text-search");
    const searchBtn = screen.getByTestId("test-search-btn");

    expect(screen.queryByPlaceholderText("Search movie")).toBe(input);
    expect(searchBar).toBeInTheDocument();
    expect(input).toBeInTheDocument();
    expect(searchBtn).toBeInTheDocument();

    expect(searchBar).toMatchSnapshot();
  });

  test("Fires action when input is filled and search button clicked", () => {
    const customStore = customStoreData(true, false, false);

    const { store } = renderWithProviders(<SearchBar />, {
      preloadedState: {
        ...customStore,
      },
    });

    const input = screen.getByTestId("input-text-search");
    const searchBtn = screen.getByTestId("test-search-btn");

    expect(store.getState().filters.search).toBe("");

    fireEvent.change(input, {
      target: {
        value: "Movie title",
      },
    });
    fireEvent.click(searchBtn);

    expect(screen.getByDisplayValue("Movie title")).toBeInTheDocument();
    expect(store.getState().filters.search).toBe("Movie title");
  });
});
