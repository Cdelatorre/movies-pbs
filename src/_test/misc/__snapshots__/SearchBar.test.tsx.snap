// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders SearchBar component correctly Renders SearchBar default component correctly 1`] = `
<div
  class="search-bar mb-5 mt-4 w-100 d-flex"
  data-testid="search-bar"
>
  <div
    class="input "
  >
    <input
      class="form-control form-text bg-dark text-light w-100"
      data-testid="input-text-search"
      name="search"
      placeholder="Search movie"
      type="text"
      value=""
    />
  </div>
  <button
    class="btn btn-success ms-3"
    data-testid="test-search-btn"
    id="search-btn"
  >
    Search
  </button>
</div>
`;
